### PR TITLE
VEY-219: Automate OS installation media setup

### PR DESCRIPTION
**Pull request by zyoung51**
Automate the process of deploying OS media (ISO) from a local/remote(http) to a local mirror and prepare the local mirror to be used as a target for network installation.

@tannoa2 @RackHD/corecommitters

### CODE REVIEW

File: data/templates/setup_iso.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        mount3 = subprocess.Popen(['awk', '{print $1}'], stdin=mount2.stdout, stdout=subprocess.PIPE)
+        isoname = mount3.communicate()[0]
+
+        iso_basename = os.path.basename(isoname).strip()
+        iso_dirname = os.path.dirname(isoname).strip()
+
+        os.makedirs(dstpath)
+        
+        syscall = '(cd "'+iso_dirname+'" && echo "'+iso_basename+'" | cpio -H newc --quiet -L -o )'
+        syscall += ' | gzip -9'
```
Comment (benbp) [inline data/templates/setup_iso.py]: I wonder if it's possible to set PIPEFAIL or something here, that way if any of these steps error, they don't get swallowed by gzip and produce a corrupted initrd image? I've had some painful debugging experiences in the past related to pipelines like this.

**Comment by benbp** [comment]
:+1:

File: data/templates/setup_iso.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    if tmpdir != '':
+        subprocess.check_call(['umount', tmpdir])
+        os.rmdir(tmpdir)
+
+def get_iso_info(fname):
+    # Access ISO to determine OS type from supported list
+    label1=subprocess.Popen(['isoinfo', '-i', fname, '-d'], stdout=subprocess.PIPE)
+    return label1.stdout.read()
+
+def do_install_os(osname, osver, src, dest):
```
Comment (yyscamper) [inline data/templates/setup_iso.py]: From your implementation for this function, seems the better name is "setup_repo"

File: data/templates/setup_iso.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import os
+import errno
+import os.path
+import tempfile
+import atexit
+import sys
+import shutil
+import re
+import urllib
+
```
Comment (yyscamper) [inline data/templates/setup_iso.py]: We need a comment on the top of file to describe how to use this script

**Comment by yyscamper** [comment]
Thanks for your work on this, this surely brings convenience for bootstrap feature user. I think you should pay attention to following item:

For CentOS 7.0, ESXi (Both 5.5 and 6.0),  the official ISO file is not workable, I have described the reason in this functional spec: https://hardware.corp.emc.com/display/gheonrack/Extend+OnRack+BootImage+API+to+support+using+external+OS+package+repository

So the ISO file need be the customized one. 

,

**Comment by yyscamper** [comment]
In Shanghai, we setup a mirror server to simplify the OS repo data copying, this mirror server connects with every stack and our develop server, the mirror server exposes data via Samba, so everyone just need to mount the mirror server to the /var/renasar/on-http/static/http folder, saving the copying time.

Meanwhile, the OS repo doesn't required be in the on-http static folder, this repo can be placed in anywhere but under the ORA's control network (currently is 172.31.128.xxx).

@jfrey had another solution for this, setup a proxy in ORA, so computer node has the ability to download from external network, the corresponding  functional spec is  https://hardware.corp.emc.com/display/gheonrack/Provide+HTTP+Proxy+in+MonoRail+to+Installation+Repos, 
@jfrey assigns this task to me, but I haven't started because now I works on some higher priority thing.

**Comment by yyscamper** [comment]
One more comment: this is just an utility script to help people setup OS repository from ISO file, I don't think it is suitable to put it under on-http repo. If we agree this is a feature for on-http, put it under data/templates is also not suitable and I prefer to implement it with Javascript. Move it to "on-tools" folder?  @benbp

**Comment by zyoung51** [comment]
@yyscamper In the short term, I want the flexibility to be able to either run it from a workflow or login and run it directly (the latter reason is why it does not belong in code).  

My preference would be to move it into on-ansible and use the Ansible workflow/tasks or ansible command line to run these types of automation tasks.

**Comment by zyoung51** [comment]
@yyscamper For ESXi, the case issue does not occur when the ISO is mounted and extracted in Linux.  I'm assuming you used 7zip, or some other tool, with a samba mount to extract the ISO, which is what introduced the issue you reference above.  Please update your wiki link to reflect this.

**Comment by yyscamper** [comment]
@zyoung51 , Yes, I used the 7-zip, it will convert the case? Actually, I tested both samba mount and non-samba mount case, my testing shows only lower case can success. If use upper case, installation seems success, but when reboot to the "success" system, it will fail and show fail to find some module.

**Comment by jfrey** [comment]
I'd like to see the proxy work done soon since it will aid the vagrant demo setup making it easier for people to try out the technology.

On Nov 12, 2015, at 6:53 AM, Felix Yuan <notifications@github.com<mailto:notifications@github.com>> wrote:


@zyoung51<https://github.com/zyoung51> , Yes, I used the 7-zip, it will convert the case? Actually, I tested both samba mount and non-samba mount case, my testing shows only lower case can success. If use upper case, installation seems success, but when reboot to the "success" system, it will fail and show fail to find some module.

**Comment by zyoung51** [comment]
@yyscamper I'm not sure how http proxy is 'another solution for this'.  This story is aimed at: "I have an ISO, how do I use the ISO to set up a location to point a netinstall/pxeboot at".  How would enabling http proxy result in an .iso being pxebootable and that .iso working in the OS install workflow as the repo= parameter?

**Comment by benbp** [comment]
Related to the discussion about moving the logic into javascript (@yyscamper @zyoung51): I thought about this a bunch when reviewing, as one alternative would be to break up these steps into more modular jobs (at the expense of more code/complexity but the benefit of greater failure tolerance/debuggability). In the end though, I think the way a lot of people will use this system is by uploading their own custom scripts in languages they are most comfortable with to data/templates, and just using the workflow engine as a way of coordinating serving these scripts. I think this is a totally reasonable use case and one that will be very common, so I'm not opposed to doing it ourselves. 

As Zach mentioned though, adding this as a set of Ansible tasks is probably the ideal solution in the long term.

**Comment by yyscamper** [comment]
@zyoung51 , if we think about from the original repo format, your solution and the http proxy can be thought as complementary. If we think about from how to let it be easier for setup the repository, both solutions are target for it.

I agree people will do get benefit from your work, my biggest concern is the place where this utility file should be in.

**Comment by yyscamper** [comment]
@zyoung51 , as you suggested, if we try to support set an iso file to repo option, just like "repo=redhat.iso", then we call your script to setup the repository under on-http static folder, that will be a good feature. If so, we can move your script to on-tasks/lib/utils/scripts folder.

**Comment by benbp** [comment]
Yeah, if a script file doesn't actually have any template variables, it probably makes sense to not call it a template.

File: data/templates/setup_iso.py (hunk 3 above)
Comment (benbp) [inline data/templates/setup_iso.py]: Or at least a `-h` flag handler, if you still intend for this to be called manually (as you say below).

**Comment by zyoung51** [comment]
This is very similar to get_smart.sh, monorail_backup/restore.sh which are not templated, but are in this same path.  I'm following the pre-established pattern here.

@benbp Right, this lets a user curl and run the command or use it with downloadUrl in a linux-command type task

@yyscampler I'm not sure why you would want python code in the javascript on-tasks library?

**Comment by benbp** [comment]
Good point. Perhaps we want to move all four (not relevant to this PR I recognize)?

**Comment by zyoung51** [comment]
@benbp Works for me.  Is on-http/data/scripts the right new home?

**Comment by yyscamper** [comment]
ahh, get-smart is an example, but not exactly. get-smart is excuted in node, node need to dowmload it from http. But your script is executed in ORA locally, doesn't involve http.

I just think it is a function for install-os task itself, so put it in on-tasks is reasonable.


Felix

-----Original Message-----

**Comment by yyscamper** [comment]
I prefer to only move this scrip in this PR, open another PR to move other scripts. We'd better do testing after each script moving, I don't want to over load Zach's work this time


Felix

-----Original Message-----

**Comment by benbp** [comment]
Oh I was misunderstanding, this is run on the server itself! This changes my opinion, clearly I wasn't reading closely enough, sorry about that.

In that case, yeah on-http has typically been our holding area for this type of stuff (case in point monorail_backup.sh), but it's not ideal and has usually been for lack of a better location at the time. I think something like on-ansible, as you say, is probably the best long term place for it. Since on-ansible itself isn't getting migrated, I think we are planning on adding another repo at some point with more deployment-type helpers from on-ansible, perhaps https://github.com/RackHD/on-tools would be a good place, I don't know). @heckj do you have any thoughts here? Co-opt on-tools for deployment helpers? Make something else? Throw it into the RackHD repo proper? @RackHD/corecommitters?

@yyscamper I wouldn't want to put this in on-tasks because I'm not sure whether it's the best idea to manage the setup of the actual server via the workflow engine, because I would assume you'd usually want to do that prior to the server running, or at least independently of it. I think something like Ansible as a driver here is a good solution, but we're also in new territory since we aren't being deployed by the proprietary pieces anymore.

TL;DR I'm fine with throwing this into on-http/data/templates for now as a short-term holding area until we have the proper repository set up.

**Comment by yyscamper** [comment]
@zyoung51 @benbp : OK, I agree to put it here as temporary solution just because we had bad example before Zach's PR (monorail-backup.sh, monorail-restore.sh and so on)  ^^.

Do you think how to follow on this, open an ODR or a new functional spec?

**Comment by amymullins** [comment]
I'll create a tracking story for the long term solution.  If it's been agreed the short term solution is acceptable, can we get this merged in?   This "short term" solution fixes a hole that was observed in the MFG Testing with OS Installation that will be very helpful for the near term milestones.

**Comment by benbp** [comment]
:+1: @amymullins

**Comment by benbp** [comment]
Needs one more

**Comment by jlongever** [comment]
:+1:
